### PR TITLE
Add date filter to activity page

### PR DESCRIPTION
**Pull request by marderer**

### CODE REVIEW

File: app/controllers/transactions_controller.rb [hunk 1]
```diff
@@ -2,7 +2,13 @@ class TransactionsController < ApplicationController
   before_action :find_transaction, only: %i[update edit]
 
   def index
-    @transactions = current_user.transactions.paginate(page: params[:page], per_page: 10)
+    @search = current_user.transactions.ransack(params[:q])
+    @transactions = @search.result(distinct: true).paginate(page: params[:page], per_page: 10)
+  end
+
+  def search
+    index
```
Comment (AleksSenkou) [inline app/controllers/transactions_controller.rb]: @marderer as I understand you need `@search` and `@transactions`  in `#index` as well as in `#serch`

For that you can add before action:
```ruby
before_action :search_for_transactions, only: %i[index search]

def index
end

def search
  render :index
end

private

def search_for_transactions
  @search = current_user.transactions.ransack(params[:q])
  @transactions = @search.result(distinct: true).paginate(page: params[:page], per_page: 10)
end
```

Comment (AleksSenkou) [reply app/controllers/transactions_controller.rb]: @marderer before making those changes be sure to add controller test to your `spec/controllers` folder. All actions from controller have to be tested with different types of data.

Comment (marderer) [reply app/controllers/transactions_controller.rb]: @AleksSenkou I added some tests for activity page. I think for CRUD actions better add where I adding this actions to activity page. ok?

File: config/routes.rb [hunk 2]
```diff
@@ -3,5 +3,9 @@
   root to: 'pages#home'
 
   get 'activity_page' => 'transactions#index'
-  resources :transactions, except: %i[index show]
+  resources :transactions, except: %i[index show] do
+    collection do
+      match 'search' => 'transactions#search', :via => %i[get post], :as => :search
```
Comment (AleksSenkou) [inline config/routes.rb]: @marderer let's use new ruby hash style `as: :search`

File: spec/controllers/transactions_controller_spec.rb [hunk 3]
```diff
@@ -0,0 +1,30 @@
+require 'rails_helper'
+
+RSpec.describe TransactionsController, type: :controller do
+  login_user
```
Comment (AleksSenkou) [inline spec/controllers/transactions_controller_spec.rb]: @marderer will be cool to wrap `login_user` into `before` block. check [best practicies](http://www.betterspecs.org/)

File: spec/controllers/transactions_controller_spec.rb [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    end
+  end
+
+  describe 'GET search' do
+    it 'has a 200 status code' do
+      get :search
+      expect(response.status).to eq(200)
+    end
+  end
+end
```
Comment (AleksSenkou) [inline spec/controllers/transactions_controller_spec.rb]: @marderer let's test CRUD actions too with different types of data. Currently, we have to test `new`, `create`, `edit`, `update` 😉

File: spec/factories/transactions.rb [hunk 5]
```diff
@@ -0,0 +1,7 @@
+FactoryBot.define do
+  factory :transaction do
+    user
+    sum 100
```
Comment (AleksSenkou) [inline spec/factories/transactions.rb]: @marderer let's use [Faker::Number.digit](https://github.com/stympy/faker/blob/master/doc/number.md)

File: spec/factories/users.rb [hunk 6]
```diff
@@ -0,0 +1,12 @@
+FactoryBot.define do
+  sequence :email do |n|
+    "test#{n}@example.com"
+  end
```
Comment (AleksSenkou) [inline spec/factories/users.rb]: 👍

File: spec/factories/users.rb [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+
+  factory :user do
+    email { generate :email }
+    password 'asdfasdf'
+    password_confirmation 'asdfasdf'
+    confirmed_at Time.current
+  end
```
Comment (AleksSenkou) [inline spec/factories/users.rb]: @marderer let's use Faker for all generated data. Purpose: check that tests pass with different generated data

File: spec/features/transactions_spec.rb [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+describe 'transactions' do
+  let(:user) { create(:user) }
+  let(:transaction) do
+    create(:transaction, user_id: user.id)
+  end
+
+  before do
+    login_as(user, scope: :user)
+  end
+end
```
Comment (AleksSenkou) [inline spec/features/transactions_spec.rb]: @marderer what do you plan to check in this file?

File: spec/support/factory_bot.rb [hunk 9]
```diff
@@ -0,0 +1,4 @@
+# spec/support/factory_bot.rb
+RSpec.configure do |config|
+  config.include FactoryBot::Syntax::Methods
+end
```
Comment (AleksSenkou) [inline spec/support/factory_bot.rb]: @marderer cool practice to move configurations to separate support files 👍

File: spec/support/factory_bot.rb [hunk 10]
```diff
@@ -0,0 +1,4 @@
+# spec/support/factory_bot.rb
```
Comment (AleksSenkou) [inline spec/support/factory_bot.rb]: @marderer do we need this comment?

File: spec/features/transactions_spec.rb [hunk 11]
```diff
@@ -0,0 +1,12 @@
+require 'rails_helper'
+
+describe 'transactions' do
+  let(:user) { create(:user) }
+  let(:transaction) do
+    create(:transaction, user_id: user.id)
```
Comment (AleksSenkou) [inline spec/features/transactions_spec.rb]: @marderer can you try to make it work in this way:
```ruby
create(:transaction, user: user)
```
?

File: app/controllers/transactions_controller.rb (hunk 1 above)
Comment (AleksSenkou) [reply app/controllers/transactions_controller.rb]: @marderer sorry, I can't get what's your question

File: spec/controllers/transactions_controller_spec.rb (hunk 3 above)
Comment (marderer) [reply spec/controllers/transactions_controller_spec.rb]: @AleksSenkou  login_user It is a ```before``` block
[discribe here](https://github.com/marderer/personal-finance/blob/d495d940efae4b9ecedcc22508ed4be7f9e43fa3/spec/support/controller_macros.rb)

File: spec/controllers/transactions_controller_spec.rb [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+  describe 'GET #search' do
+    it 'has a 200 status code' do
+      get :search
+      expect(response.status).to eq(200)
+    end
+  end
+
+  describe 'POST #create' do
+    context 'when valid' do
+      it 'create transaction' do
```
Comment (AleksSenkou) [inline spec/controllers/transactions_controller_spec.rb]: @marderer `it 'creates ...'`

File: spec/controllers/transactions_controller_spec.rb [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+        put :update, params: {
+          id: transaction.id,
+          transaction: attributes_for(:transaction,
+            amount: amount,
+            date: date)
+        }
+        transaction.reload
+        expect(transaction.amount).to eq(amount)
+        expect(transaction.date).to eq(date)
+      end
```
Comment (AleksSenkou) [inline spec/controllers/transactions_controller_spec.rb]: @marderer this `it` block is overloaded. Let's give it a breath 😄 
Check notes from [here](https://github.com/sinpai/datarockets-finance-app/pull/7#discussion_r173711209) and [here](https://github.com/sinpai/datarockets-finance-app/pull/9#discussion_r173713486)

File: spec/controllers/transactions_controller_spec.rb [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+        put :update, params: {
+          id: transaction.id,
+          transaction: attributes_for(:transaction,
+            amount: nil,
+            date: nil)
+        }
+        transaction.reload
+        expect(transaction.amount).to eq(init_amount)
+        expect(transaction.date).to eq(init_date)
+      end
```
Comment (AleksSenkou) [inline spec/controllers/transactions_controller_spec.rb]: @marderer the same comment as above

**Pull request merged**